### DISCUSSION TITLE
Question on conditional / guarded transitions

### DISCUSSION THREAD

**Discussion by dsathyakumar** [statelyai/xstate#2262] [Q&A]
I have 2 questions:
1. How do we declare an array of conditional guards for a given target?
2. Given a conditional transition bound by `cond` guard & if it fails, how & where do we update the reason for failure?

Asking this cos, due to (1) i'm dumping all validations (say 10 different validation rules) of a piece of data (on various parameters) into a single guard.
So, if this guard failed, i'd have no way to know which of those 10 rules failed. 

I could add a `choose` action to the `actions`, but that would re-run all those validation rules again. 
I'm talking of something like this:
```js
{
   target: "inActive"
   cond: "didValidationFail",
   actions: choose([
     {actions: ["updateNotValidFileObj"], cond: "isValidFileObj"]},
     {actions: ["updateSizeLimitExceeded"], cond: "isCorrectSize"]}
     {actions: ["updateFileTypeIncorrect"], cond: "isRightFileType"]}
     {actions: ["updateRecordCountIncorrect"], cond: "isRightRecordCount"]},
     {actions: ["updateGenericError"]}
  ])
}
```
In the above, choose **actions** would execute if `didValidationFail` was **true**.
So in `didValidationFail` , i'm running all the validations once to decide if the machine can transition from current state to `inActive` (if the validations failed).
But, to know which validation rule failed, i'd be re-running them inside a `choose` action individually, to update which rule failed.

Is there a better way around this?

**Comment by davidkpiano** [comment]
> How do we declare an array of conditional guards for a given target?

You can't currently; this is coming to v5 though!

> Given a conditional transition bound by `cond` guard & if it fails, how & where do we update the reason for failure?

One way to model this traditionally is by having chained "choice states", which are transient states that only have guarded `always` transitions, such as:

```ts
// transient state
checkName: {
  always: [
    { target: 'checkAge', cond: 'nameValid' },
    { target: 'fail', actions: assign({ reason: ... }) }
  ]
}
```

Visually, you can think about it like this:

<img width="476" alt="CleanShot 2021-06-01 at 22 52 44@2x" src="https://user-images.githubusercontent.com/1093738/120417169-6aa35b00-c32c-11eb-9713-67dc4a1a21e7.png">

**Reply by dsathyakumar** [reply]
Thanks @davidkpiano . I did something of this sort -> start of with an event whose `actions` does an array of `choose` actions, thereby being able to run validations and update if the validation failed, while emitting a validation failed event, so it transitions to a "failed" state.
